Pre-release QA before v1.1.0

# Wave 1.1.0 pre-release QA

## Test results

- **Unit tests**

All pass

- **Visual regression tests**

As expected, only minor (barely noticeable) changes were introduced: left-side cells in markdown tables has no left padding anymore and links are little bit bolder.

![image](https://github.com/h2oai/wave/assets/23740173/db424e9c-cf47-4e1d-93ff-9f28712088ff)

<img width="1451" alt="image" src="https://github.com/h2oai/wave/assets/23740173/ea7039a4-ce45-4914-84ca-e820acf07c90">

## Manual testing 

Components changed since 1.0.0 were tested.

ui.audio_annotator ❌
ui.chatbot ✅
ui.card_menu ✅
ui.inline ✅
ui.image_annotator ✅
ui.link ✅
ui.markdown ✅
ui.notification_bar ✅
ui.picker ✅
ui.spin_box ✅
ui.stats ✅
ui.table ✅
ui.tall_stats  ✅
ui.textbox ✅
ui.toolbar ❌

### Audio annotator

1. Timeline works inconsistently across browsers, different files and file formats on Safari and Firefox:

While trying [aac sample](https://filesamples.com/samples/audio/aac/sample3.aac), cursor stucks few seconds before end on Safari and jumps to the end on Firefox:

https://github.com/h2oai/wave/assets/23740173/6380f24f-4220-48aa-864e-cd6d0a3f1ca9

It gets even more interesting for [flac](https://filesamples.com/samples/audio/flac/sample3.flac) files. The music is still playing, but the cursor is jumping on one place until the music ends. This was reproduced only for Safari:

https://github.com/h2oai/wave/assets/23740173/47af5498-73ba-427f-af47-6d72f216692a

However when trying [different aac file](https://dl.espressif.com/dl/audio/ff-16b-2c-44100hz.aac), issue persists for Safari only.

[Ogg](https://file-examples.com/wp-content/storage/2017/11/file_example_OOG_1MG.ogg) file was working for Chrome and Firefox, however on Safari it couldn't be decoded:

<img width="1415" alt="image" src="https://github.com/h2oai/wave/assets/23740173/77c3622e-9283-44e5-90b7-7884ee098d2b">

The [mp4 ](https://download.samplelib.com/mp4/sample-5s.mp4) file cannot be played on Safari:

https://github.com/h2oai/wave/assets/23740173/eb9a4545-0c99-45c6-a881-faa64653ecb7

For [m4a](https://filesamples.com/samples/audio/m4a/Symphony%20No.6%20(1st%20movement).m4a), wav and mp3 files I was not able to reproduce any of the issues.

2. Audio annotator not properly loaded when being on different OSX desktop which is not currently displayed on any of the monitors.

https://github.com/h2oai/wave/assets/23740173/fcbf7aaa-92c1-4548-9f4d-61c8829a902f

@mturoci we had this kind of issue in past with ui.plot

3. Both wave server and app crashed while playing around with audio annotator in multiple browsers:
- Logs from waved server:
```
...
2024/02/26 13:39:20 # {"addr":"127.0.0.1:57534","route":"/900f344a-5b9c-4500-a7be-5b5fd1c484b6","t":"ui_add"}
2024/02/26 13:39:20 * /900f344a-5b9c-4500-a7be-5b5fd1c484b6 {"d":[{"k":"example","d":{"view":"form","box":"1 1 -1 -1","items":[{"audio_annotator":{"name":"annotator","title":"Drag to annotate","path":"/_f/8a239a20-8c90-42a5-82dd-a3c3b1441452/audio.mp3","tags":[{"name":"f","label":"Flute","color":"$blue"},{"name":"d","label":"Drum","color":"$brown"}]}},{"button":{"name":"submit","label":"Submit","primary":true}}]}}]}
2024/02/26 13:39:20 # {"path":"/_f/8a239a20-8c90-42a5-82dd-a3c3b1441452/audio.mp3","t":"file_download"}
2024/02/26 13:39:22 # {"addr":"127.0.0.1:57521","t":"ui_drop"}
panic: concurrent write to websocket connection

goroutine 280 [running]:
github.com/gorilla/websocket.(*messageWriter).flushFrame(0x14000696e78, 0x1, {0x10306f278?, 0x140002e2f00?, 0x376de3a3ef6?})
        /Users/mmihok/go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:610 +0x46c
github.com/gorilla/websocket.(*Conn).WriteMessage(0x140000f58c0, 0x484393777f?, {0x10306f278, 0x0, 0x0})
        /Users/mmihok/go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:763 +0xf4
github.com/h2oai/wave.(*Client).flush(0x140001564d0)
        /Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/client.go:249 +0x26c
created by github.com/h2oai/wave.(*SocketServer).ServeHTTP
        /Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/socket.go:99 +0x604
exit status 2
make: *** [run] Error 1
mmihok@Mareks-MacBook-Pro wave % 
```

- Logs from `audio_annotator.py` app:
```
INFO:     127.0.0.1:57535 - "POST /disconnect HTTP/1.1" 200 OK
Unhandled exception
Traceback (most recent call last):
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 69, in map_httpcore_exceptions
    yield
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 373, in handle_async_request
    resp = await self._pool.handle_async_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection_pool.py", line 216, in handle_async_request
    raise exc from None
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection_pool.py", line 196, in handle_async_request
    response = await connection.handle_async_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection.py", line 101, in handle_async_request
    return await self._connection.handle_async_request(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/http11.py", line 143, in handle_async_request
    raise exc
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/http11.py", line 113, in handle_async_request
    ) = await self._receive_response_headers(**kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/http11.py", line 186, in _receive_response_headers
    event = await self._receive_event(timeout=timeout)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/http11.py", line 224, in _receive_event
    data = await self._network_stream.read(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_backends/anyio.py", line 31, in read
    with map_exceptions(exc_map):
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ReadError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/h2o_wave/h2o_wave/server.py", line 360, in _process
    await self._handle(q)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/examples/audio_annotator.py", line 35, in serve
    await q.page.save()
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/h2o_wave/h2o_wave/core.py", line 634, in save
    await self.site._save(self.url, p)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/h2o_wave/h2o_wave/core.py", line 887, in _save
    res = await self._http.patch(_rebase(_config.hub_address, url), content=patch)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1966, in patch
    return await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1574, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1661, in send
    response = await self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1689, in _send_handling_auth
    response = await self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1726, in _send_handling_redirects
    response = await self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1763, in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 372, in handle_async_request
    with map_httpcore_exceptions():
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 86, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadError
Failed transmitting unhandled exception
Traceback (most recent call last):
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 69, in map_httpcore_exceptions
    yield
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 373, in handle_async_request
    resp = await self._pool.handle_async_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection_pool.py", line 216, in handle_async_request
    raise exc from None
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection_pool.py", line 196, in handle_async_request
    response = await connection.handle_async_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection.py", line 99, in handle_async_request
    raise exc
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection.py", line 76, in handle_async_request
    stream = await self._connect(request)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_async/connection.py", line 122, in _connect
    stream = await self._network_backend.connect_tcp(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_backends/auto.py", line 30, in connect_tcp
    return await self._backend.connect_tcp(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_backends/anyio.py", line 112, in connect_tcp
    with map_exceptions(exc_map):
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc) from exc
httpcore.ConnectError: All connection attempts failed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/h2o_wave/h2o_wave/server.py", line 372, in _process
    await q.page.save()
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/h2o_wave/h2o_wave/core.py", line 634, in save
    await self.site._save(self.url, p)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/h2o_wave/h2o_wave/core.py", line 887, in _save
    res = await self._http.patch(_rebase(_config.hub_address, url), content=patch)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1966, in patch
    return await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1574, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1661, in send
    response = await self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1689, in _send_handling_auth
    response = await self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1726, in _send_handling_redirects
    response = await self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_client.py", line 1763, in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 372, in handle_async_request
    with map_httpcore_exceptions():
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/mmihok/Documents/h2o-ai/projects/wave/repo/wave/py/venv/lib/python3.11/site-packages/httpx/_transports/default.py", line 86, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ConnectError: All connection attempts failed
```

I was not able to reproduce the exact cause, it always happens in different scenario:

https://github.com/h2oai/wave/assets/23740173/6c58adef-fe7f-4896-b9b8-136e21e1549b


### Toolbar

Image is not downloaded directly if `download=True` if path is external link - https://wave.h2o.ai/img/logo.svg
I advise updating `toolbar.py` with the same path as in `link.py` example - http://localhost:10101/assets/brand/h2o-wave-b&w.png


https://github.com/h2oai/wave/assets/23740173/6768a909-af8f-470a-9a0b-0d59e61e3587

--- a/py/examples/toolbar.py
+++ b/py/examples/toolbar.py
@@ -18,7 +18,7 @@ async def serve(q: Q):
             ),
             ui.command(name='upload', label='Upload', icon='Upload'),
             ui.command(name='share', label='Share', icon='Share'),
-            ui.command(name='download', label='Download', icon='Download', path='https://wave.h2o.ai/img/logo.svg', download=True),
+            ui.command(name='download', label='Download', icon='Download', path='http://localhost:10101/assets/brand/h2o-wave-b&w.png', download=True),
         ],
         secondary_items=[
             ui.command(name='tile', caption='Grid View', icon='Tiles'),

--- a/ui/src/audio_annotator.test.tsx
+++ b/ui/src/audio_annotator.test.tsx
@@ -395,7 +395,7 @@ describe('AudioAnnotator.tsx', () => {
       expect(wave.args[name]).toMatchObject([items[0], { ...items[1], start: end - moveOffset, end: start }])
     })
 
-    it.only('Does not remove active tag when leaving and getting back into the zoom rectangle during dragging', async () => {
+    it('Does not remove active tag when leaving and getting back into the zoom rectangle during dragging', async () => {
       const { container } = render(<XAudioAnnotator model={model} />)
       await waitForComponentLoad()
       const canvasEl = container.querySelector('canvas')!

--- a/ui/src/audio_annotator.test.tsx
+++ b/ui/src/audio_annotator.test.tsx
@@ -395,6 +395,19 @@ describe('AudioAnnotator.tsx', () => {
       expect(wave.args[name]).toMatchObject([items[0], { ...items[1], start: end - moveOffset, end: start }])
     })
 
+    it.only('Does not remove active tag when leaving and getting back into the zoom rectangle during dragging', async () => {
+      const { container } = render(<XAudioAnnotator model={model} />)
+      await waitForComponentLoad()
+      const canvasEl = container.querySelector('canvas')!
+
+      expect(container.querySelector('[class*=activeTag]')).toBeInTheDocument()
+
+      fireEvent.mouseDown(canvasEl, { clientX: 30, clientY: 10, buttons: 1 })
+      fireEvent.mouseLeave(canvasEl)
+      fireEvent.click(canvasEl, { clientX: 31, clientY: 20, buttons: 1 })
+
+      expect(container.querySelector('[class*=activeTag]')).toBeInTheDocument()
+    })
   })
 
   describe('Tooltip', () => {

--- a/ui/src/parts/range_annotator.tsx
+++ b/ui/src/parts/range_annotator.tsx
@@ -366,7 +366,7 @@ const
           : 'pointer'
 
         if (!currDrawnAnnotation.current || !action) {
-          if (intersected && intersected.tag !== activeTag) setActiveTag(intersected.tag)
+          if (intersected && intersected.tag !== activeTag && intersected.tag !== '') setActiveTag(intersected.tag)
           if (intersected) focusAnnotation(intersected)
           return
         }

--- a/ui/src/parts/range_annotator.tsx
+++ b/ui/src/parts/range_annotator.tsx
@@ -408,7 +408,13 @@ const
 
     React.useEffect(() => {
       window.addEventListener('resize', init)
-      return () => window.removeEventListener('resize', init)
+      // HACK: Fix canvas having default dimensions of 300x150 if app is loaded when browser tab is not visible.
+      const onVisibilityChange = () => { if (canvasRef.current && canvasRef.current.width === 300) setTimeout(init, 60) }
+      window.addEventListener("visibilitychange", onVisibilityChange)
+      return () => {
+        window.removeEventListener('resize', init)
+        window.removeEventListener("visibilitychange", onVisibilityChange)
+      }
     }, [init])
 
     React.useEffect(() => redrawAnnotations(), [annotations, redrawAnnotations])

--- a/ui/src/parts/waveform.tsx
+++ b/ui/src/parts/waveform.tsx
@@ -50,8 +50,13 @@ export const Waveform = React.memo(({ color, data }: Props) => {
   React.useLayoutEffect(() => {
     updateDimensions()
     const onResize = debounce(1000, updateDimensions)
+    const onVisibilityChange = () => setTimeout(updateDimensions, 60)
     window.addEventListener('resize', onResize)
-    return () => window.removeEventListener('resize', onResize)
+    window.addEventListener('visibilitychange', onVisibilityChange)
+    return () => {
+      window.removeEventListener('resize', onResize)
+      window.removeEventListener('visibilitychange', onVisibilityChange)
+    }
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [])
 

--- a/website/widgets/form/link.md
+++ b/website/widgets/form/link.md
@@ -56,7 +56,7 @@ q.page['example'] = ui.form_card(box='1 1 2 2', items=[
 
 If you want to allow your users to download a file, use the `download` attribute.
 
-The `download` attribute will only start the download if the path points to a location on the same origin (if your app is running on <http://localhost:10101>, the `path` either starts with `/` or `http://localhost:10101`) or has a `Content-disposition` response header set to `attachment`. See [the docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-download).
+The `download` attribute will only start the download if the path points to a location on the same origin (wave server).
 
 ```py
 q.page['example'] = ui.form_card(box='1 1 2 2', items=[

